### PR TITLE
Improve replication

### PR DESCRIPTION
**Pull request by axlbonnet**
Correct bugs linked to replication and healing plugin.

### CODE REVIEW

File: src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java [hunk 1]
```diff
@@ -174,23 +215,39 @@ public void run() {
...
+                    ") and received (" + returnedJobIds.size() + ") jobs" +
+                    " when checking status");
+        }
+        StringBuilder missingJobIds = new StringBuilder();
+        for (String sentJobId : sentJobIds) {
+            if ( !returnedJobIds.contains(sentJobId)) {
+                missingJobIds.append(sentJobId);
+                missingJobIds.append(" ");
+            }
+        }
```
Comment (camarasu) [inline src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java]: As discussed, in order to simplify this, we could only retrieve the missingJobIds if the sizes are not equal

File: src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java [hunk 2]
```diff
@@ -148,6 +172,8 @@ public void run() {
                                     case Stalled:
                                         job.setStatus(GaswStatus.STALLED);
                                         break;
+                                    case Completed:
+                                        logger.info("Dirac Monitor: job \"" + job.getId() + "\" has status \"" + status + "\"");
```
Comment (camarasu) [inline src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java]: Shouldn't case "Completed" be handled the same as "Done" ? I would say that these two Dirac statuses are equivalent from VIP's point of view

File: src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java [hunk 3]
```diff
@@ -315,7 +379,32 @@ protected void resume(Job job) {
...
+            for (JobMinorStatus minorStatus : factory.getJobMinorStatusDAO().getExecutionMinorStatus(job.getId())) {
+
+                switch (minorStatus.getStatus()) {
+
+                    case Application:
+                        job.setRunning(minorStatus.getDate());
+                        break;
+                    case Outputs:
+                        job.setUpload(minorStatus.getDate());
+                }
```
Comment (camarasu) [inline src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java]: Why handle only job.setRunning and job.setUpload and why break after Application ? Don't we want to set the date for all minor statuses ?

File: src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java (hunk 2 above)
Comment (axlbonnet) [reply src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java]: I think that failed jobs can also go through the "COMPLETED" status.
Also I think we should wait for the "DONE" status for the output and error files to be available.

File: src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java (hunk 3 above)
Comment (axlbonnet) [reply src/main/java/fr/insalyon/creatis/gasw/plugin/executor/dirac/execution/DiracMonitor.java]: I kept the same behavior as in GaswOutputParser::parseNonStdOut.
The idea is that "queued" and "download" times should have known through dirac status changes, but, without minor statuses, "running" and "upload" times are lost when there is no output file available (killed and stalled jobs).

**Comment by camarasu** [comment]
Great job !

**Pull request merged**